subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/checkstyle/CheckstylePluginIntegrationTest.groovy

```diff
@@ -55,4 +66,174 @@ class CheckstylePluginIntegrationTest extends WellBehavedPluginTest {
+    @See("https://checkstyle.sourceforge.io/config_system_properties.html#Enable_External_DTD_load")
+    def "can use enable_external_dtd_load feature on task to override extension value for a task"() {
+        given:
+        buildFile """
+            apply plugin: 'checkstyle'
+
+            ${mavenCentralRepository()}
+
+            checkstyle {
+                enableExternalDtdLoad = false
+            }
+
+            tasks.withType(Checkstyle).configureEach {
+                enableExternalDtdLoad = true
+            }
+        """
+
+        file('src/main/java/org/sample/MyClass.java') << multilineJavaClass()
+        file('config/checkstyle/checkstyle-common.xml') << checkStyleCommonXml()
+        file('config/checkstyle/checkstyle.xml') << checkStyleMainXml()
+
+        when:
+        fails 'checkstyleMain'
+
+        then:
+        result.assertHasErrorOutput("Checkstyle rule violations were found. See the report at:")
+        result.assertHasErrorOutput("Checkstyle files with violations: 1")
+        result.assertHasErrorOutput("Checkstyle violations by severity: [error:2]")
+        file("build/reports/checkstyle/main.xml").assertContents(containsClass("org.sample.MyClass"))
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/21624")
+    @See("https://checkstyle.sourceforge.io/config_system_properties.html#Enable_External_DTD_load")
+    def "if use enable_external_dtd_load feature NOT enabled, error if feature used in rules XML"() {
+        given:
+        buildFile """
+            apply plugin: 'checkstyle'
+
+            ${mavenCentralRepository()}
+
+            checkstyle {
+                enableExternalDtdLoad = false
+            }
+        """
+
+        file('src/main/java/org/sample/MyClass.java') << multilineJavaClass()
+        file('config/checkstyle/checkstyle-common.xml') << checkStyleCommonXml()
+        file('config/checkstyle/checkstyle.xml') << checkStyleMainXml()
+
+        when:
+        fails 'checkstyleMain'
+
+        then:
+        failedDueToXmlDTDProcessingError()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/21624")
+    @See("https://checkstyle.sourceforge.io/config_system_properties.html#Enable_External_DTD_load")
+    def "enable_external_dtd_load feature NOT enabled by default"() {
+        given:buildFile """
+            apply plugin: 'checkstyle'
+
+            ${mavenCentralRepository()}
+        """
+
+        file('src/main/java/org/sample/MyClass.java') << multilineJavaClass()
+        file('config/checkstyle/checkstyle-common.xml') << checkStyleCommonXml()
+        file('config/checkstyle/checkstyle.xml') << checkStyleMainXml()
+
+        when:
+        fails 'checkstyleMain'
+
+        then:
+        failedDueToXmlDTDProcessingError()
+    }
+
+    private failedDueToXmlDTDProcessingError() {
+        result.assertHasErrorOutput("A failure occurred while executing org.gradle.api.plugins.quality.internal.CheckstyleAction")
```

Comment:
You should use `failure.assertHasCause(...)` here instead.  `assertHasErrorOutput` just looks for this message _somewhere_ in the error output, but `failure.assertHasCause(...)` requires that it shows up in a particular part of the build output.